### ISSUE TITLE
Classloader confusion when passing an externally imported class into a script's binding

### ISSUE THREAD

**Issue by enigma0Z** [jenkinsci/JenkinsPipelineUnit#202]
**Expected:** When I pass an object into the binding of a pipeline run in the test harness, it uses it without complaining about class loader issues

**Actual:** When calling another method (defined in Jenkins shared library) that uses this object, unless it's a `java.lang` object like `String` or a groovy object like `GStringImpl`, it complains about differing class loaders.  Example code below.

**Use Case**
In our current pipeline shared library, we have a variety of data classes defined to represent our deployment structure, which applications are built, and when they are built where they go.  There is a need to pass these objects into pipelines in order to validate that they have correct behavior.  It'd be better to define the test data in the test suite than to define it somewhere within our actual shared library so that the script suite can load it.

-----

**MyDataClass.groovy**
```
Class MyDataClass {
    String value
}
```

**MyMethodClass.groovy**
```groovy
import MyDataClass

class MyMethodClass {
    void doStuff(MyDataClass myDataClass) {
        println(myDataClass.value)
    }
}
```

**MyPipeline.groovy**
```groovy
import MyMethodClass

MyMethodClass myMethodClass = new MyMethodClass()

myMethodClass.doStuff(myDataClass) // <---- Fails here
```

**MyTest.groovy**
```groovy
import MyDataClass

class MyTest extends BasePipelineTest {
    @Test
    void doTest() {
        helper.setUp()
        binding.setVariable('myDataClass', new MyDataClass('foo'))
        runScript('MyPipeline.groovy')
    }
}
```

This (above) fails where noted due to mismatched classloader in between object being passed in and class definition... with an error like this:
```
    groovy.lang.MissingMethodException: No signature of method: MyMethodClass.doStuff() is applicable for argument types: (MyDataClass) values: [...whatever]
    Possible solutions: <lists the method signature>
    The following classes appear as argument class and as parameter class, but are defined by different class loader:
    MyDataClass (defined by 'sun.misc.Launcher$AppClassLoader@2a139a55' and 'groovy.lang.GroovyClassLoader$InnerLoader@5ea25338')
    If one of the method suggestions matches the method you wanted to call, 
    then check your class loader setup.
```

**Comment by stchar** [comment]
Hi @enigma0Z 
I have a question about the example
```
doStuff(MyClass myClass) { ...
```  
is it a typo? should it be `MyDataClass` instead of `MyClass`

**Comment by enigma0Z** [comment]
Whoops, yes that's a typo.  I initially called the class `MyClass` when I was writing up the example but changed it and must have missed that one.

**Comment by stchar** [comment]
ok, I see. This is a tricky one. JPU cretes own class loader to parse pipeline scripts and to load library classes, 
see [InterceptingGCL](https://github.com/jenkinsci/JenkinsPipelineUnit/blob/master/src/main/groovy/com/lesfurets/jenkins/unit/InterceptingGCL.groovy). 

I can see few options: 
1. Create a base class for your data classes put it in a dedicated jar and load it as dependency for both junit and shared library via `@Grape`.
```
class MyDataClass extends MyBaseClass {
}
```
Then doStuff signature could be changed to 
```
doStuff(MyBaseClass myClass) {} 
```
2. Use [LibClassLoader](https://github.com/jenkinsci/JenkinsPipelineUnit/blob/master/src/main/groovy/com/lesfurets/jenkins/unit/LibClassLoader.groovy) which is used to support dynamic library definition
Create an utility pipeline script
```
// utils.jenkinsfile
return library('library_with_data_classes')
```
Update test
```
void doTest() {
        helper.setUp()
        def libcl = runScript('utils.jenkinsfile')
        binding.setVariable('myDataClass', libcl.MyDataClass.new('foo'))
        runScript('MyPipeline.groovy')
    }
```

3. You can try to update [PipelineTestHelper](https://github.com/jenkinsci/JenkinsPipelineUnit/blob/master/src/main/groovy/com/lesfurets/jenkins/unit/PipelineTestHelper.groovy):
* store reference to InterceptingGCL 
* add a public method to load any class using [parseClass](https://github.com/jenkinsci/JenkinsPipelineUnit/blob/master/src/main/groovy/com/lesfurets/jenkins/unit/InterceptingGCL.groovy#L36) 
```
void doTest() {
        helper.setUp()
        helper.parseClass('src/test/MyDataClass.groovy').newInstance('foo')
        binding.setVariable('myDataClass', helper.parseClass('src/MyDataClass.groovy').newInstance('foo'))
        runScript('MyPipeline.groovy')
    }
```

**Comment by stchar** [comment]
@enigma0Z  looks like #193 may fix your issue

**Comment by stchar** [comment]
@enigma0Z ,  could you please try version `1.4`

**Comment by stchar** [comment]
@enigma0Z ,  could you please try version `1.4`

**Comment by enigma0Z** [comment]
Thanks, I will check it this week.

**Comment by enigma0Z** [comment]
Tried `1.4`, but at first glance at least the issue is still present.  It might be due to how some of our pipeline classes are being loaded into the test harness.  I want to try a few more scenarios and see which ones if any are improved.

**Comment by enigma0Z** [comment]
After further investigation, it looks like #193 / version `1.4` does not fix the issue.  Still receiving the same class loader complaint.

**Comment by stchar** [comment]
@enigma0Z, It's hard give you the right answer. I don't know how groovy class loader works, the entrance point to start digging in could be 
https://github.com/jenkinsci/JenkinsPipelineUnit/blob/master/src/main/groovy/com/lesfurets/jenkins/unit/InterceptingGCL.groovy

Please also take a look at #223

**Comment by enigma0Z** [comment]
We ended up working around this issue by pre-loading objects into the binding context before they were accessed by the library.

**Comment by stefano-m** [comment]
See also https://github.com/jenkinsci/JenkinsPipelineUnit/issues/381#issuecomment-1189097421

> I have been bitten by this too, after much debugging, I did not find an easy way to modify the class loader.
> 
> **However**, setting
> 
> ```groovy
> helper.libLoader.preloadLibraryClasses = false
> ```
> 
> did fix both this issue and #202
> 
> The `preloadLibraryClasses` setting is [somewhat documented in the README](https://github.com/jenkinsci/JenkinsPipelineUnit/blob/1caecc45ea253c715398ee0ed2b09f35e066f3ce/README.md#library-global-variables-accepting-library-class-instances-as-arguments-troubleshooting) but this specific case is not listed.

**Comment by qtruong-jnpr** [comment]
@enigma0Z I am new to this lib, can you show me sample code how to get over this issue. I jump into the same rabbit hole but can't not use the solution from @stefano-m 
Thank you!
